blockchain/types/src/block.rs

```diff
@@ -123,14 +135,19 @@ impl Block {
     ///   block*
     /// * `root_element` - The root element for membership proofs
     /// * `block_contents` - Contents of the block.
+    /// * `timestamp` - The timestamp of the block in ms since Unix epoch.
+    ///   should be 0 for block versions 3 and below.
     pub fn new(
         version: BlockVersion,
         parent_id: &BlockID,
         index: BlockIndex,
         cumulative_txo_count: u64,
         root_element: &TxOutMembershipElement,
         block_contents: &BlockContents,
+        timestamp: u64,
     ) -> Self {
+        assert!(timestamp == 0 && version.timestamps_are_supported());
+
```

Comment:
I'll leave it to you @eranrund on keeping the assert.

I'm a bit scarred from being in code bases where people were too aggressive on asserts and it prevented one from quickly getting started.
At the same time having the error surface as soon as possible in the pipeline is nice.